etl/tests/common/bigquery.rs

```diff
@@ -0,0 +1,371 @@
+        }
+    }
+
+    /// Takes ownership of the BigQuery client, leaving [`None`] in its place.
+    ///
+    /// Used during cleanup to move the client out for async dataset deletion.
+    fn take_client(&mut self) -> Option<Client> {
+        match self {
+            Self::Emulated { client, .. } => client.take(),
+            Self::Real { client, .. } => client.take(),
+        }
+    }
+}
+
+impl Drop for BigQueryDatabase {
+    /// Cleans up the test dataset when the database instance is dropped.
+    ///
+    /// Automatically deletes the BigQuery dataset and all its tables to
+    /// ensure test isolation and prevent resource leaks.
+    fn drop(&mut self) {
+        // We take out the client since during destruction we know that the struct won't be used
+        // anymore.
+        let Some(client) = self.take_client() else {
+            return;
+        };
+
+        // To use `block_in_place,` we need a multithreaded runtime since when a blocking
+        // task is issued, the runtime will offload existing tasks to another worker.
+        tokio::task::block_in_place(move || {
+            Handle::current().block_on(async move {
+                destroy_bigquery(&client, self.project_id(), self.dataset_id()).await;
+            });
+        });
+    }
+}
+
+/// Creates a new BigQuery dataset for testing.
+///
+/// Sets up a fresh dataset in the specified project that will be used
+/// for all table operations during the test.
+async fn initialize_bigquery(client: &Client, project_id: &str, dataset_id: &str) {
+    let dataset = Dataset::new(project_id, dataset_id);
+    client.dataset().create(dataset).await.unwrap();
+}
+
+/// Deletes a BigQuery dataset and all its contents.
+///
+/// Removes the test dataset and all tables within it to clean up
+/// resources after testing.
+async fn destroy_bigquery(client: &Client, project_id: &str, dataset_id: &str) {
+    client
+        .dataset()
+        .delete(project_id, dataset_id, true)
+        .await
+        .unwrap();
+}
+
+/// Sets up a BigQuery database connection for testing.
+///
+/// Automatically detects whether to use real Google Cloud BigQuery or the local
+/// emulator based on environment variables. If `TESTS_BIGQUERY_SA_KEY_PATH` is
+/// set, connects to real BigQuery; otherwise uses the emulated version.
+///
+/// Creates a fresh dataset for test isolation and configures mock authentication
+/// when using the emulator.
+pub async fn setup_bigquery_connection() -> BigQueryDatabase {
+    let sa_key_path = std::env::var(BIGQUERY_SA_KEY_PATH_ENV_NAME);
+
+    match sa_key_path {
+        Ok(sa_key_path) => BigQueryDatabase::new_real(sa_key_path).await,
+        Err(_) => {
+            let google_auth = GoogleAuthMock::start().await;
+            google_auth.mock_token(1).await;
+
+            let sa_key = serde_json::to_string_pretty(&mock_sa_key(&google_auth.uri())).unwrap();
+
+            BigQueryDatabase::new_emulated(google_auth, sa_key).await
+        }
+    }
```

Comment:
For now I decided to add support for both emulated and real BigQuery, however, we might want to remove the emulated one if we see that it becomes to expensive to test against a real BigQuery instance.